### DISCUSSION TITLE
Example ex-prt-mp7-p01  Problem running

### DISCUSSION THREAD

**Discussion by JJHelly** [modflowpy/flopy#2407] [Q&A] [answered]
Aloha.

I'm running MF6 and not quite sure what Modpath 7 is but it seems to be a class within flopy.  Not sure.  Running on OSX Sequoia.

I'm making good progress in developing a structured grid for a volcano and want to test with an example.  I found ex-prt-mp7-p01 and it seems like a challenging, but relevant, example case for my application.  However, it immediately fails with what appears to be a flopy problem.  Before I spend inordinate amounts of time trying to figure it out,  I wonder is someone can tell me what it's complaining about and how to correct it (see below)?  Also, if not, then perhaps where I should start debugging?
Mahalo.
J.

> Building models for ex-prt-mp7-p01
> ---------------------------------------------------------------------------
> FlopyException                            Traceback (most recent call last)
> Cell In[46], line 2
>       1 # We are now ready to run the example problem. Subproblems 1A and 1B are solved by a single MODFLOW 6 run and a single MODPATH 7 run, so they are included under one "scenario". Each of the two subproblems is represented by its own particle release package (for MODFLOW 6) or particle group (for MODPATH 7).
> ----> 2 scenario(silent=True)
> 
> Cell In[45], line 2, in scenario(silent)
>       1 def scenario(silent=False):
> ----> 2     gwfsim, prtsim, mp7 = build_models(sim_name)
>       3     if write:
>       4         write_models(gwfsim, prtsim, mp7, silent=silent)
> 
> Cell In[42], line 137, in build_models(example_name)
>     134 # Instantiate the MODFLOW 6 prt particle release point (prp) package for example 1A,
>     135 # first converting MP7 particle release configurations to PRT format.
>     136 releasepts_1a = list(mp7_particle_data_1a.to_prp(prt.modelgrid))
> --> 137 flopy.mf6.ModflowPrtprp(
>     138     prt,
>     139     pname="prp1a",
>     140     filename=f"{prt_name}_1a.prp",
>     141     nreleasepts=len(releasepts_1a),
>     142     packagedata=releasepts_1a,
>     143     perioddata={0: ["FIRST"]},
>     144     exit_solve_tolerance=1e-5,
>     145     extend_tracking=True,
>     146 )
>     148 # Instantiate the MODFLOW 6 prt particle release point (prp) package for example 1B,
>     149 # first converting MP7 particle release configurations to PRT format.
>     150 releasepts_1b = list(mp7_particle_data_1b.to_prp(prt.modelgrid))
> 
> File /Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/flopy/mf6/modflow/mfprtprp.py:587, in ModflowPrtprp.__init__(self, model, loading_package, boundnames, print_input, dev_exit_solve_method, exit_solve_tolerance, local_z, track_filerecord, trackcsv_filerecord, stoptime, stoptraveltime, stop_at_weak_sink, istopzone, drape, release_timesrecord, release_timesfilerecord, dev_forceternary, nreleasepts, packagedata, perioddata, filename, pname, **kwargs)
>     561 def __init__(
>     562     self,
>     563     model,
>    (...)
>     585     **kwargs,
>     586 ):
> --> 587     super().__init__(
>     588         model, "prp", filename, pname, loading_package, **kwargs
>     589     )
>     591     # set up variables
>     592     self.boundnames = self.build_mfdata("boundnames", boundnames)
> 
> File /Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/flopy/mf6/mfpackage.py:1868, in MFPackage.__init__(self, parent, package_type, filename, pname, loading_package, **kwargs)
>    1864 kwargs_str = ", ".join(kwargs.keys())
>    1865 excpt_str = (
>    1866     f'Extraneous kwargs "{kwargs_str}" provided to MFPackage.'
>    1867 )
> -> 1868 raise FlopyException(excpt_str)
> 
> FlopyException: Extraneous kwargs "extend_tracking" provided to MFPackage. ()

![image](https://github.com/user-attachments/assets/c29ef593-80c0-43be-b9c3-9029d9bb8387)

**Comment by wpbonelli** [comment]
Hi @JJHelly long story short, make sure you have the latest flopy (3.9.1) and the latest MF6 (6.6.0) and this should resolve. A few points below to add some context.

[MODPATH 7](https://www.usgs.gov/software/modpath-particle-tracking-model-modflow) (repo [here](https://github.com/MODFLOW-USGS/modpath-v7)) is a separate particle tracking program in the MODFLOW ecosystem. PRT brings particle tracking into MF6 proper and generalizes the tracking algorithm to unstructured grids. Flopy has a class providing an interface to MP7, indeed. AFAIK, MP7 is the most commonly used particle tracker with MF6 at this time, hence the effort in this example to demonstrate how to build equivalent MP7 and MF6 PRT models and/or migrate existing MP7 models to PRT.

Much of the `flopy.mf6` namespace is an auto-generated interface layer for MF6. When we make an MF6 release, we also release a new flopy in which this module is synchronized to the new MF6. Flopy 3.9.1 is paired with MF6.6.0. Both came out last week. In MF6.6.0 we added an option called `extend_tracking` to the PRT Particle Release Point (PRP) package. Likewise flopy 3.9.1 now has a corresponding parameter in the `flopy.mf6.Prtprp` class initializer. The [MF6.6.0 release notes](https://github.com/MODFLOW-USGS/modflow6/releases/tag/6.6.0) have more info on this change.

Flopy provides a way to sync its mf6 module without upgrading. See [here](https://flopy.readthedocs.io/en/stable/md/generate_classes.html) for more info. If you anticipate (or are already) working with the MF6 nightly build, this is the way to go &mdash; make sure to use `--ref develop` to sync with the very latest. If you are working only with stable MF6 releases, upgrading flopy is the easiest way.

Let me know if anything above isn't clear, I have left out some detail for brevity.

It is good to hear someone else is trying out PRT. Cheers.

**Comment by JJHelly** [comment]
Aloha.

Mahalo for the generous, and quick, response.  I didn't realize I had stumbled onto such a cutting edge example.  That's great.  I'm making pretty good progress with MF6 so your efforts are very much appreciated.

I've worked out a method for extracting lava flows from the USGS lithology database and am developing some code to build multi-layer grids, wedding cake style, of volcanoes based on the history of surficial lava flows that can eventually be extrapolated with core data to greater depths.  At some point it might make a good example for others since it provides for terrain models to be employed in other settings; especially where road cuts provide stratigraphic data.

This is probably TMI but I mention it in the interest of contributing something back.

J.
------------------------------------------------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 25, 2024, at 16:59, wpbonelli ***@***.***> wrote:



Hi @JJHelly<https://urldefense.com/v3/__https://github.com/JJHelly__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkOxVf1_3V$> long story short, make sure you have the latest flopy (3.9.1) and the latest MF6 (6.6.0) and this should resolve. A few points below to add some context.

MODPATH 7<https://urldefense.com/v3/__https://www.usgs.gov/software/modpath-particle-tracking-model-modflow__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO7G5IqUv$> (repo here<https://urldefense.com/v3/__https://github.com/MODFLOW-USGS/modpath-v7__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO2JBh4th$>) is a separate particle tracking program in the MODFLOW ecosystem. PRT brings particle tracking into MF6 proper and generalizes the tracking algorithm to unstructured grids. Flopy has a class providing an interface to MP7, indeed. AFAIK, MP7 is the most commonly used particle tracker with MF6 at this time, hence the effort in this example to demonstrate how to build equivalent MP7 and MF6 PRT models and/or migrate existing MP7 models to PRT.

Much of the flopy.mf6 namespace is an auto-generated interface layer for MF6. When we make an MF6 release, we also release a new flopy in which this module is synchronized to the new MF6. Flopy 3.9.1 is paired with MF6.6.0. Both came out last week. In MF6.6.0 we added an option called extend_tracking to the PRT Particle Release Point (PRP) package. Likewise flopy 3.9.1 now has a corresponding parameter in the flopy.mf6.Prtprp class initializer. The MF6.6.0 release notes<https://urldefense.com/v3/__https://github.com/MODFLOW-USGS/modflow6/releases/tag/6.6.0__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO8UAiRUC$> have more info on this change.

Flopy provides a way to sync its mf6 module without upgrading. See here<https://urldefense.com/v3/__https://flopy.readthedocs.io/en/stable/md/generate_classes.html__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO0w-8D2T$> for more info. If you anticipate (or are already) working with the MF6 nightly build, this is the way to go — make sure to use --ref develop to sync with the very latest. If you are working only with stable MF6 releases, upgrading flopy is the easiest way.

Let me know if anything above isn't clear, I have left out some detail for brevity.

It is good to hear someone else is trying out PRT. Cheers.

**Comment by JJHelly** [comment]
Aloha.

Trying to follow you advice to use the git version but run into this:


***@***.*** modflow6 % export LDFLAGS="$LDFLAGS -Wl,-ld_classic"

***@***.*** modflow6 % meson setup build

The Meson build system

Version: 1.6.1

Source dir: /Volumes/Pegasus32R6/Projects-300/Project-USGS-Modeling/modflow6

Build dir: /Volumes/Pegasus32R6/Projects-300/Project-USGS-Modeling/modflow6/build

Build type: native build

Project name: MODFLOW 6

Project version: 6.7.0.dev0


meson.build:1:0: ERROR: Unable to get gcc pre-processor defines:

Compiler stdout:


-----

Compiler stderr:

gfortran: fatal error: cannot execute 'cc1': execvp: No such file or directory

compilation terminated.


Don't know why cc1 is missing as I've just build modpath yesterday.  I can't find it anywhere and seems it should have come with gcc13 which, BTW, comes with gfortran-mp-13, not gfortran.  Perhaps something is not configured correctly but if you have any thoughts, I'd appreciate it.

J.

----------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 25, 2024, at 16:59, wpbonelli ***@***.***> wrote:



Hi @JJHelly<https://urldefense.com/v3/__https://github.com/JJHelly__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkOxVf1_3V$> long story short, make sure you have the latest flopy (3.9.1) and the latest MF6 (6.6.0) and this should resolve. A few points below to add some context.

MODPATH 7<https://urldefense.com/v3/__https://www.usgs.gov/software/modpath-particle-tracking-model-modflow__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO7G5IqUv$> (repo here<https://urldefense.com/v3/__https://github.com/MODFLOW-USGS/modpath-v7__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO2JBh4th$>) is a separate particle tracking program in the MODFLOW ecosystem. PRT brings particle tracking into MF6 proper and generalizes the tracking algorithm to unstructured grids. Flopy has a class providing an interface to MP7, indeed. AFAIK, MP7 is the most commonly used particle tracker with MF6 at this time, hence the effort in this example to demonstrate how to build equivalent MP7 and MF6 PRT models and/or migrate existing MP7 models to PRT.

Much of the flopy.mf6 namespace is an auto-generated interface layer for MF6. When we make an MF6 release, we also release a new flopy in which this module is synchronized to the new MF6. Flopy 3.9.1 is paired with MF6.6.0. Both came out last week. In MF6.6.0 we added an option called extend_tracking to the PRT Particle Release Point (PRP) package. Likewise flopy 3.9.1 now has a corresponding parameter in the flopy.mf6.Prtprp class initializer. The MF6.6.0 release notes<https://urldefense.com/v3/__https://github.com/MODFLOW-USGS/modflow6/releases/tag/6.6.0__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO8UAiRUC$> have more info on this change.

Flopy provides a way to sync its mf6 module without upgrading. See here<https://urldefense.com/v3/__https://flopy.readthedocs.io/en/stable/md/generate_classes.html__;!!Mih3wA!Ct4_PkcEtwedz2ByJxEFziYl44trMGP1ap9uznn6_xbBIIaGURcmEJc5FUC75GZ9iDhMKGrn_D-kXflkO0w-8D2T$> for more info. If you anticipate (or are already) working with the MF6 nightly build, this is the way to go — make sure to use --ref develop to sync with the very latest. If you are working only with stable MF6 releases, upgrading flopy is the easiest way.

Let me know if anything above isn't clear, I have left out some detail for brevity.

It is good to hear someone else is trying out PRT. Cheers.

**Reply by wpbonelli** [reply]
I'm guessing `gfortran-mp-xx` is a macports idiosyncrasy? It looks like meson can't find the correct gfortran. It's probably looking for `gfortran` by default. Maybe try setting `FC=gfortran-mp-13` or linking that executable to `gfortran`?

For what it's worth, we only run automated tests with the homebrew toolchain, not the one from macports, so can't guarantee the latter is sound. But if you've just used it to build modpath, that's surely not the problem..

Note there are instructions to build MF6 [here](https://modflow6.readthedocs.io/en/latest/_dev/developer.html#building), in case you have not already found them.

If you don't want/need to build from source, MF6 can be [installed with a command provided by flopy](https://flopy.readthedocs.io/en/stable/md/get_modflow.html). For instance, to get the latest official release:

```sh
get-modflow : --repo modflow6
```

For the bleeding edge, swap `--repo modflow6-nightly-build`.

Thanks for sharing your use case. Very interested to see that take shape. I'd imagine such an example would be welcome.

**Comment by JJHelly** [comment]
Aloha.

Mahalo for the advice.  gfortran-mp-13 is a macports-derived code but I don't know if it's GNU convention or not.  I was going to dig into the compiler universe to see if I could figure out the problem.  I simply copied the above to ~/bin/gfortran as that worked for me some long time ago for mf6; so, I don't know why it's different now but apparently so.

In any event, I have installed 6.6 from binary and updated flopy and the example runs fine although there is some warning somewhere that I haven't dug into yet.

I'fl keep you apprised as this model comes along.

J.
------------------------------------------------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 27, 2024, at 03:55, wpbonelli ***@***.***> wrote:



I'm guessing gfortran-mp-xx is a macports idiosyncrasy? It looks like meson can't find the correct gfortran. It's probably looking for gfortran by default. Maybe try setting FC=gfortran-mp-13 or linking that executable to gfortran?

For what it's worth, we only run automated tests with the homebrew toolchain, not the one from macports, so can't guarantee the latter is sound. But if you've just used it to build modpath, that's surely not the problem..

Note there are instructions to build MF6 here<https://urldefense.com/v3/__https://modflow6.readthedocs.io/en/latest/_dev/developer.html*building__;Iw!!Mih3wA!HJGY2dvMZYH9w0OSilPw1-6Er-dx1E723V7EDeS7sO4WcwgIG_bM_YFfbDge2VWRQ4TZp0jxxZCIBMUvHC62IfMB$>, in case you have not already found them.

If you don't want/need to build from source, MF6 can be installed with a command provided by flopy<https://urldefense.com/v3/__https://flopy.readthedocs.io/en/stable/md/get_modflow.html__;!!Mih3wA!HJGY2dvMZYH9w0OSilPw1-6Er-dx1E723V7EDeS7sO4WcwgIG_bM_YFfbDge2VWRQ4TZp0jxxZCIBMUvHKMpQ_Bn$>. For instance, to get the latest official release:

get-modflow : --repo modflow6

For the bleeding edge, swap --repo modflow6-nightly-build.

Thanks for sharing your use case. Very interested to see that take shape. I'd imagine such an example would be welcome.

**Comment by JJHelly** [comment]
FWIW, I did follow those build instructions but the problem seems to be with the compiler.  Any chance in finding out what compiler USGS folks use to produce the binaries on OSX ARM?

J

------------------------------------------------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 27, 2024, at 03:55, wpbonelli ***@***.***> wrote:



I'm guessing gfortran-mp-xx is a macports idiosyncrasy? It looks like meson can't find the correct gfortran. It's probably looking for gfortran by default. Maybe try setting FC=gfortran-mp-13 or linking that executable to gfortran?

For what it's worth, we only run automated tests with the homebrew toolchain, not the one from macports, so can't guarantee the latter is sound. But if you've just used it to build modpath, that's surely not the problem..

Note there are instructions to build MF6 here<https://urldefense.com/v3/__https://modflow6.readthedocs.io/en/latest/_dev/developer.html*building__;Iw!!Mih3wA!HJGY2dvMZYH9w0OSilPw1-6Er-dx1E723V7EDeS7sO4WcwgIG_bM_YFfbDge2VWRQ4TZp0jxxZCIBMUvHC62IfMB$>, in case you have not already found them.

If you don't want/need to build from source, MF6 can be installed with a command provided by flopy<https://urldefense.com/v3/__https://flopy.readthedocs.io/en/stable/md/get_modflow.html__;!!Mih3wA!HJGY2dvMZYH9w0OSilPw1-6Er-dx1E723V7EDeS7sO4WcwgIG_bM_YFfbDge2VWRQ4TZp0jxxZCIBMUvHKMpQ_Bn$>. For instance, to get the latest official release:

get-modflow : --repo modflow6

For the bleeding edge, swap --repo modflow6-nightly-build.

Thanks for sharing your use case. Very interested to see that take shape. I'd imagine such an example would be welcome.

**Comment by christianlangevin** [comment]
Hi @JJHelly, I use the homebrew install of gfortran on my ARM Mac.  Information is below.

```
$ which gfortran
/opt/homebrew/bin/gfortran
$ gfortran --version
GNU Fortran (Homebrew GCC 14.1.0_2) 14.1.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Reply by wpbonelli** [reply]
We also use homebrew gcc for the ARM mac distribution, only difference being version 13 not 14.

**Comment by JJHelly** [comment]
Mahalo.  Appreciated.
J.
----------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 27, 2024, at 09:59, langevin-usgs ***@***.***> wrote:



Hi @JJHelly<https://urldefense.com/v3/__https://github.com/JJHelly__;!!Mih3wA!G43PWjjrQQaLqHSHPoWQMKqzSoRvUZgqcuscqrb2l-NdlXaJq6WoNjB18V6K5TpGRBqw2YpSvYXWRzEh8_KmyUnH$>, I use the homebrew install of gfortran on my ARM Mac. Information is below.

$ which gfortran
/opt/homebrew/bin/gfortran
$ gfortran --version
GNU Fortran (Homebrew GCC 14.1.0_2) 14.1.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Comment by JJHelly** [comment, marked as answer]
Got it.  This seems to be a good incantation.

I switched to the macports gcc-mp-14 and gfortran-mp-14 and got past that problem.  Just linked to them via ~/bin.   I suspect the problem has something to do with not locating the libraries if I just copy the compilers and not the libraries (duh).

I'll send a note when I have the compile done later (hopefully).
J.
----------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 27, 2024, at 10:06, wpbonelli ***@***.***> wrote:



We also use homebrew gcc for the ARM mac distribution, only difference being version 13 not 14.

**Comment by JJHelly** [comment]
Bingo.


***@***.*** src % ./mf6 --version


mf6: 6.7.0.dev0 (preliminary) 12/19/2024


***@***.*** src %


----------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 27, 2024, at 10:06, wpbonelli ***@***.***> wrote:



We also use homebrew gcc for the ARM mac distribution, only difference being version 13 not 14.

**Reply by wpbonelli** [reply]
Excellent. Please let us know if you notice anything needing attention in the developer docs as those are mostly internal-use up to now.

**Comment by JJHelly** [comment]
Will do.
J.
------------------------------------------------------------------------------------------------------------------------------------------------------------
John Helly / San Diego Supercomputer Center / Scripps Institution of Oceanography
https://www.sdsc.edu/~hellyj / 808 205 9882 / 760 8408660

On Dec 28, 2024, at 02:51, wpbonelli ***@***.***> wrote:



Excellent. Please let us know if you notice anything needing attention in the developer docs as those are mostly internal-use up to now.
